tools/rewriter/GenCallAsm.cpp

```diff
@@ -280,6 +286,300 @@ static std::string sig_string(const CAbiSignature &sig,
+      add_comment_line(aw, "Set rdi to the compartment's return value memory");
+      add_asm_line(aw, "movq %rsp, %rdi");
+      // The new return value is 8 bytes above the bottom of the stack so we need
+      // to add 8 to rdi
+      add_asm_line(aw, "addq $8, %rdi");
+    }
+
+    // Insert 8 bytes to align the stack to 16 bytes if necessary.
+    if (stack_alignment != 0) {
+      assert(stack_alignment == 8);
+      add_asm_line(aw, "subq $8, %rsp");
+    }
+
+    // Copy stack args to target stack
+    if (stack_arg_count > 0) {
+      // Set rax to the caller's stack so we can copy the stack args to the
+      // compartment's stack.
+      add_comment_line(
+          aw, "Copy stack arguments from the caller's stack to the compartment");
+      // Load addr of top of caller compartment's stack into rax, clobbering r12
+      std::string expr = emit_load_sp_offset(aw, caller_pkey, "r12"s);
+      add_asm_line(aw, "movq %" + expr + ", %rax");
+      // This is effectively a memcpy of size `stack_arg_size` from the caller's
+      // stack to the compartment's
+      for (int i = 0; i < stack_arg_size; i += 8) {
+        // We must take the preserved registers we pushed on the caller's stack
+        // into account (including rbp) when determining the location of the stack
+        // args
+        size_t caller_stack_size =
+            stack_arg_size + ((preserved_registers.size() + 1) * 8);
+        // The index into the caller's stack is backwards since pushq will copy to
+        // the compartment's stack from the highest addresses to the lowest.
+        add_asm_line(aw,
+                     "pushq " + std::to_string(caller_stack_size - i) + "(%rax)");
+      }
+    }
+  } else if (arch == Arch::Aarch64) {
+    // TODO ARM arguments
+    llvm::errs() << "TODO arguments not implemented on ARM\n";
+  }
+}
+
+static void emit_load_fn_ptr(AsmWriter &aw, Arch arch) {
+  if (arch == Arch::X86) {
+    add_asm_line(aw, "movq ia2_fn_ptr@GOTPCREL(%rip), %r12");
+    add_asm_line(aw, "movq (%r12), %r12");
+  } else if (arch == Arch::Aarch64) {
+    // TODO ARM function pointer save
+    llvm::errs() << "TODO indirect calls not implemented on ARM\n";
+  }
+}
+
+static void emit_set_pkru(AsmWriter &aw, uint32_t target_pkey, Arch arch) {
+  if (arch == Arch::X86) {
+    // Change pkru to the compartment's value
+    add_comment_line(aw, "Set PKRU to the compartment's value");
+    // wrpkru requires zeroing rcx and rdx, but they may have arguments so use r10
+    // and r11 as scratch registers
+    add_asm_line(aw, "movq %rcx, %r10");
+    add_asm_line(aw, "movq %rdx, %r11");
+    emit_wrpkru(aw, target_pkey);
+    add_asm_line(aw, "movq %r10, %rcx");
+    add_asm_line(aw, "movq %r11, %rdx");
+  } else if (arch == Arch::Aarch64) {
+    // set X18 to the pointer key (compartment number left-shifted 56 bits)
+    add_asm_line(aw, "mov x18, #" + std::to_string(target_pkey & 0xF));
+    add_asm_line(aw, "lsl x18, x18, #56");
```

Comment:
you can set x18 with one instruction with something like `movz x18, #0x0100, LSL 48". This sets x18 to `0x0100 << 48` so we just need to modify the upper byte in the 16-bit immediate field.